classes/loginflow/authcode.php

```diff
@@ -618,11 +618,27 @@ protected function handlelogin(string $oidcuniqid, array $authparams, array $tok
             $username = trim(\core_text::strtolower($username));
             $tokenrec = $this->createtoken($oidcuniqid, $username, $authparams, $tokenparams, $idtoken, 0, $originalupn);
             $userinfo = $this->get_userinfo($username);
-            if (!$CFG->allowaccountssameemail && array_key_exists('email', $userinfo)) {
-                $existinguserparams = ['email' => $userinfo['email'], 'mnethostid' => $CFG->mnet_localhost_id];
-            } else {
-                $existinguserparams = ['username' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
+            
+            $matchingmethod = get_config('auth_oidc', 'subjectmapping');
+            switch ($matchingmethod) {
+                case 'username':
+                    $existinguserparams = ['username' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
+                    break;
+                case 'idnumber':
+                    // In this case, we are treating the "username" from OIDC as the idnumber
+                    $existinguserparams = ['idnumber' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
+                    break;
+                case 'email':
+                    // We can only match on email if allowaccountssameemail is off.
+                    if (!$CFG->allowaccountssameemail && array_key_exists('email', $userinfo)) {
+                        $existinguserparams = ['email' => $userinfo['email'], 'mnethostid' => $CFG->mnet_localhost_id];
+                    } else {
+                        // Fallback to username.
+                        $existinguserparams = ['username' => $username, 'mnethostid' => $CFG->mnet_localhost_id];
```

Comment:
Is there a way we can refactor this so that we don't duplicate line 637 with line 625

Potentially move the `allowaccountsameemail` check outside of the switch, and if it fails just change the `$matchingmethod` to username